records/2023-03-28.secret-management.md

```diff
@@ -0,0 +1,189 @@
+| Self Hosted Vault   | Yes               | Yes                    | Yes                       | Yes                       | Yes                         | Yes            | No              | No           | Free         | High             | Medium              |
+| AWS Secrets Manager | Yes               | No (One per account)   | No                        | Yes                       | No                          | No             | Yes             | No           | Free         | Medium           | Low                 |
+| AWS Parameter Store | Yes               | No (One per account)   | No                        | No                        | No                          | No             | Yes             | No           | Free         | Medium           | Free                |
+
+## Additional considerations
+
+### Vault - Common
+
+#### Kubernetes Integration
+
+[Kubernetes Integration is done via the vault secret injector combined with the vault sidecar.](https://developer.hashicorp.com/vault/tutorials/kubernetes/kubernetes-external-vault)
+
+The secret injector is scoped to a service account within kubernetes, and creates a JWT token that is used to authenticate with Vault. The sidecar is a container that runs in parallel with the target pod (e.g. admin or api). The sidecar periodically checks Vault for updates to the secrets and automatically injects the secrets into the target pod. When the secret is updated, the pod does not need to be refreshed, however the code must be written in such a way that it will automatically check for new values for this to be effective. The integration will require modifications to the Kustomize (or Helm) deployment for each application that needs to use it. These modifications are referenced just once. 
+
+
+#### AWS First Party Services Integration
+
+[AWS Lamba Integration is done via the Vault Lambda extension.](https://developer.hashicorp.com/vault/tutorials/app-integration/aws-lambda)
+
+The vault lambda extension must be added to the target code application and deployed within the docker image. Additional logic must be added to the container to actually call vault and leverage the secrets. The extension works by assuming a specific AWS IAM role, and using that to authenticate against Vault. The set up for this is rather complex and convoluted as it requires Vault to also use aws as an authentication back end (So vault authenticates against aws, and aws authenticates against vault). This can (and should) be configured through terraform, and thus once it is done, it will be easy to make changes and run against new environmments. As with Kubernetes, the secret can be updated in vault and it will automatically propagate to the function. **To be confirmed**
```

Comment:
Ah that answers a question I was wondering about. I guess we could start by restarting pods and gradually write necessary abstractions to retrieve the secrets within the code to have no downtime, if we want to go this way.

Reply:
Yeah exactly - and there will definitely be some R&D required for this to figure out exactly how it will work with Lambdas